File: androidApp/src/main/java/at/jodlidev/esmira/views/main/RewardView.kt
package at.jodlidev.esmira.views.main

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.widget.Toast
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import at.jodlidev.esmira.*
import at.jodlidev.esmira.R
import at.jodlidev.esmira.sharedCode.NativeLink
import at.jodlidev.esmira.sharedCode.data_structure.ErrorBox
import at.jodlidev.esmira.sharedCode.data_structure.Questionnaire
import at.jodlidev.esmira.sharedCode.data_structure.Study
import at.jodlidev.esmira.views.DefaultButtonIconAbove
import at.jodlidev.esmira.views.DialogButton
import at.jodlidev.esmira.views.ESMiraDialog
import at.jodlidev.esmira.views.inputViews.TextElView

/**
 * Created by JodliDev on 28.02.2023.
 */
@Composable
fun RewardView(
	getStudy: () ->Study,
	goBack: () -> Unit
) {
	val context = LocalContext.current
	val study = getStudy()
    val getCode = remember { mutableStateOf(study.hasCachedRewardCode())}
    val requestRewardCode = {getCode.value = true}
	val error = remember { mutableStateOf("") }
	val fulfilledQuestionnaires = remember {
		mutableMapOf<Long, Boolean>()
	}

	val rewardCode = produceState(initialValue = "", getCode.value) {
        if(getCode.value) {
            study.getRewardCode(
                onError = { error.value = it },
                onSuccess = { rewardInfo ->
                    ErrorBox.log(
                        "Reward",
                        "Received reward response (errorMessage: ${rewardInfo.errorMessage}, errorCode: ${rewardInfo.errorCode}, fulfilledQuestionnaires: ${rewardInfo.fulfilledQuestionnaires})"
                    )
                    when (rewardInfo.errorCode) {
                        Study.REWARD_SUCCESS ->
                            value = rewardInfo.code

                        Study.REWARD_ERROR_UNFULFILLED_REWARD_CONDITIONS -> {
                            error.value =
                                context.getString(R.string.error_reward_conditions_not_met)
                            fulfilledQuestionnaires.putAll(rewardInfo.fulfilledQuestionnaires)
                        }

                        Study.REWARD_ERROR_ALREADY_GENERATED ->
                            error.value = context.getString(R.string.error_already_generated)

                        else -> error.value = rewardInfo.errorMessage
                    }
                }
            )
        } else {
            fulfilledQuestionnaires.putAll(study.getRewardFulfillmentLocal())
            value = ""
        }
	}
	DefaultScaffoldView(title = stringResource(R.string.rewards), goBack = goBack) {
		Column(
			modifier = Modifier
				.padding(all = 20.dp)
				.fillMaxWidth()
		) {
			val untilActive = study.daysUntilRewardsAreActive()
			if(!study.enableRewardSystem || untilActive != 0) {
				val resources = LocalContext.current.resources
				RewardDefaultView(
					study,
					resources.getQuantityString(R.plurals.info_reward_is_not_active_yet, untilActive, untilActive),
					HashMap()
                ) {}
            } else if (study.enableRewardCalculation) {
                if(error.value.isNotEmpty()) {
                    RewardDefaultView(study, error.value, fulfilledQuestionnaires, requestRewardCode)
                } else if(rewardCode.value.isNotEmpty()) {
                    RewardCodeView(study, rewardCode.value)
                } else if(getCode.value) {
                    RewardLoadingView()
                } else {
                    RewardDefaultView(study, "", fulfilledQuestionnaires, requestRewardCode)
                }
            }
			else {
				if(error.value.isNotEmpty())
					RewardDefaultView(study, error.value, fulfilledQuestionnaires, requestRewardCode)
				else if(rewardCode.value.isEmpty())
					RewardLoadingView()
				else
					RewardCodeView(study, rewardCode.value)
			}
		}
	}
}

@Composable
fun RewardCodeView(study: Study, rewardCode: String) {
	val showRewardCodeExplanation = remember { mutableStateOf(false) }
	val context = LocalContext.current
	if(showRewardCodeExplanation.value) {
		ESMiraDialog(
			onDismissRequest = { showRewardCodeExplanation.value = false },
			title = stringResource(R.string.what_for),
			confirmButtonLabel = stringResource(R.string.close),
			onConfirmRequest = { showRewardCodeExplanation.value = false },
		) {
			Text(stringResource(R.string.reward_code_description))
		}
	}
	Row(
		verticalAlignment = Alignment.CenterVertically
	) {
		Text(stringResource(R.string.colon_reward_code_header))
		Spacer(modifier = Modifier.size(10.dp))
		DialogButton(stringResource(id = R.string.what_for), onClick = { showRewardCodeExplanation.value = true })
	}
	
	Spacer(modifier = Modifier.size(10.dp))
	
	Column(
		modifier = Modifier
			.fillMaxWidth(),
		horizontalAlignment = Alignment.CenterHorizontally
	) {
		TextButton(
			onClick = {
			val clipBoard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
				val clipData = ClipData.newPlainText("label", rewardCode)
				clipBoard.setPrimaryClip(clipData)
				Toast.makeText(context, context.getString(R.string.android_info_copied_x_to_clipboard, rewardCode), Toast.LENGTH_SHORT).show()
			},
			colors = ButtonDefaults.textButtonColors(contentColor = MaterialTheme.colorScheme.tertiary)
		) {
			Text(rewardCode, fontSize = MaterialTheme.typography.headlineLarge.fontSize, fontWeight = FontWeight.Bold)
			Spacer(Modifier.size(ButtonDefaults.IconSpacing))
			Icon(Icons.Default.ContentCopy,
				contentDescription = "copy",
				modifier = Modifier.size(ButtonDefaults.IconSize)
			)
		}
		
		Spacer(modifier = Modifier.size(30.dp))
		
		Row(
			modifier = Modifier.fillMaxWidth(),
			horizontalArrangement = Arrangement.SpaceEvenly
		) {
			if(study.contactEmail.isNotEmpty()) {
				DefaultButtonIconAbove(
					icon = Icons.Default.Email,
					text = stringResource(R.string.reward_code_send_email),
					modifier = Modifier.weight(0.45F),
					onClick = {
						val intent = Intent(Intent.ACTION_SEND)
						val emailContent = study.rewardEmailContent.ifEmpty { context.getString(R.string.reward_code_content) }
						intent.putExtra(Intent.EXTRA_EMAIL, arrayOf(study.contactEmail))
						intent.putExtra(Intent.EXTRA_SUBJECT, context.getString(R.string.android_email_study_subject, study.title))
						intent.putExtra(Intent.EXTRA_TEXT, emailContent.replace("[[CODE]]", rewardCode))
						intent.type = "text/plain"
						try {
							context.startActivity(intent)
						} catch(e: Exception) {
							Toast.makeText(context, R.string.error_no_email_client, Toast.LENGTH_LONG).show()
						}
					}
				)
				
				Spacer(modifier = Modifier.weight(0.1F))
			}
			DefaultButtonIconAbove(
				icon = Icons.Default.Share,
				text = stringResource(R.string.reward_code_share),
				modifier = Modifier.weight(0.45F),
				onClick = {
					val intent = Intent(Intent.ACTION_SEND)
					intent.putExtra(Intent.EXTRA_TEXT, rewardCode)
					intent.type = "text/plain"
					try {
						context.startActivity(intent)
					} catch(e: Exception) {
						Toast.makeText(context, R.string.error_no_email_client, Toast.LENGTH_LONG).show()
					}
				}
			)
		}
	}
	if(study.rewardInstructions.isNotEmpty()) {
		Spacer(modifier = Modifier.size(20.dp))
		Column(modifier = Modifier.verticalScroll(rememberScrollState())) {
			HtmlHandler.HtmlText(study.rewardInstructions, modifier = Modifier.fillMaxWidth())
		}
	}
    if(study.enableRewardCalculation) {
        Spacer(modifier = Modifier.size(20.dp))
        Column(modifier = Modifier.verticalScroll(rememberScrollState())) {
            if(study.rewardCalculationInfo.isNotEmpty()) {
                HtmlHandler.HtmlText(
                    study.rewardCalculationInfo,
                    modifier = Modifier.fillMaxWidth()
                )
            }
            Text(stringResource(R.string.reward_final_amount, study.cachedRewardAmount))
        }
    }
}


@Composable
fun RewardLoadingView() {
	Column(
		modifier = Modifier.fillMaxSize(),
		verticalArrangement = Arrangement.Center,
		horizontalAlignment = Alignment.CenterHorizontally
	) {
		CircularProgressIndicator()
	}
}

@Composable
fun RewardDefaultView(study: Study, error: String, fulfilledQuestionnaires: Map<Long, Boolean>, requestRewardCode: () -> Unit) {
    val untilActive = study.daysUntilRewardsAreActive()
    val rewardAvailable = untilActive <= 0
    val showDialog = remember { mutableStateOf(false) }
    val canRequest = rewardAvailable && fulfilledQuestionnaires.all{(_, fulfilled) -> fulfilled}

    Column(
		modifier = Modifier.fillMaxWidth(),
		horizontalAlignment = Alignment.CenterHorizontally
	) {
		Text(error, textAlign = TextAlign.Center)
	}
	
	if(fulfilledQuestionnaires.isNotEmpty() && study.questionnaires.isNotEmpty()) {
		Spacer(modifier = Modifier.size(30.dp))
		Text(stringResource(id = R.string.error_reward_questionnaires_not_finished))
		Spacer(modifier = Modifier.size(10.dp))
		
		val availableQuestionnaires = study.questionnaires.filter {
			// always display unfulfilled questionnaires, for user feedback in case of older servers not marking inaccessible questionnaires as fulfilled
			questionnaire: Questionnaire -> questionnaire.limitToGroup == 0 || questionnaire.limitToGroup == study.group || fulfilledQuestionnaires[questionnaire.internalId] != true
		}
		
		availableQuestionnaires.forEach { questionnaire ->
			Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
				Text("${questionnaire.title}:", modifier = Modifier.weight(1F))
				val modifier = Modifier.width(50.dp)
				if(fulfilledQuestionnaires[questionnaire.internalId] == true)
					Icon(Icons.Default.CheckCircle, "true", tint = colorGreen, modifier = modifier)
				else
					Icon(Icons.Default.Cancel, "false", tint = colorRed, modifier = modifier)
			}
		}
//		LazyVerticalGrid(
//			columns = GridCells.Fixed(2),
//			horizontalArrangement = Arrangement.Center,
//			verticalArrangement = Arrangement.Center
//		) {
//			study.questionnaires.forEach { questionnaire ->
//				item {
//					Text("${questionnaire.title}:")
//				}
//				item {
//					Box(contentAlignment = Alignment.Center, modifier = Modifier.fillMaxHeight()) {
//						if(fulfilledQuestionnaires[questionnaire.internalId] == true)
//							Icon(Icons.Default.CheckCircle, "true", tint = colorGreen)
//						else
//							Icon(Icons.Default.Cancel, "false", tint = colorRed)
//					}
//				}
//			}
//		}
	}

    if(study.enableRewardCalculation) {
        HtmlHandler.HtmlText(study.rewardCalculationInfo)
        val rewardAmount = study.getRewardAmount()
        Column(
            modifier = Modifier.fillMaxWidth(),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            if(study.rewardCalculationInfo.isNotEmpty()) {
                HtmlHandler.HtmlText(study.rewardCalculationInfo, modifier = Modifier.fillMaxWidth())
            }
            Text(
                stringResource(
                    R.string.reward_current_amount,
                    "%.2f".format(study.getRewardAmount())
                )
            )
        }
    }

    Column(
        modifier = Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        if(!rewardAvailable) {
            val resources = LocalContext.current.resources
            Text(
                resources.getQuantityString(
                    R.plurals.info_reward_is_not_active_yet,
                    untilActive,
                    untilActive
                )
            )
        }
            Button(enabled = canRequest, onClick = {
            showDialog.value = true
        }) {
            Text(stringResource(R.string.reward_code_request))
        }
    }

    if(showDialog.value){
        ESMiraDialog(confirmButtonLabel = stringResource(R.string.reward_code_request_action),
            onConfirmRequest = { requestRewardCode() },
            title = stringResource(R.string.reward_code_request),
            dismissButtonLabel = stringResource(R.string.cancel),
            onDismissRequest = { showDialog.value = false}
            ) {
            if(study.enableRewardCalculation) {
                Text(stringResource(R.string.reward_code_request_info_calculation))
            } else {
                Text(stringResource(R.string.reward_code_request_info))
            }
        }
    }

}

@Preview
@Composable
fun PreviewNotActiveYet() {
	ESMiraSurface {
		val study = Study.newInstance("", "", """{"id":1, "enableRewardSystem": true, "rewardVisibleAfterDays": 1}""")
		study.joinedTimestamp = NativeLink.getNowMillis()
		RewardDefaultView(study, "Preview error message", HashMap()) {}
	}
}

@Preview
@Composable
fun ErrorWithFulfilledQuestionnairesPreview() {
	ESMiraSurface {
		val study = Study.newInstance(
			"",
			"",
			"""{"id":1, "enableRewardSystem": true, "questionnaires": [{"title": "This is a very long title for a questionnaire that hopefully has a break", "internalId": 1}, {"title": "questionnaire 2", "internalId": 2}]}"""
		)
		Column(modifier = Modifier.width(400.dp)) {
			RewardDefaultView(study, "Preview error message", mapOf(Pair(1L, true), Pair(2L, false))) {}
		}
	}
}

File: iosApp/ESMira/Views/RewardView.swift
//
//  Created by JodliDev on 12.10.22.
//

import SwiftUI
import sharedCode


struct RewardView: View {
	@EnvironmentObject var appState: AppState
	@State var study: Study
	
	@State private var showShareSheet = false
	@State private var error: String = ""
	@State private var rewardCode: String = ""
	@State private var fulfilledQuestionnaires: [KotlinLong : KotlinBoolean] = [:]
	@State private var getCode: Bool = false
	@State private var showRequestConfirmation: Bool = false
	
	@State private var showAlert: Bool = false
	@State private var alertView: () -> Alert = { Alert(title: Text(""))}
	
	private var canRequest: Bool {
		let untilActive = self.study.daysUntilRewardsAreActive()
		let rewardAvailable = untilActive <= 0
		let allFulfilled = fulfilledQuestionnaires.values.allSatisfy {$0.boolValue}
		return rewardAvailable && allFulfilled
	}
	
	init(study: Study) {
		self.study = study
	}
	
	private func getDefaultView(error: String) -> some View {
		VStack(alignment: .leading) {
			if(!error.isEmpty) {
				Text(error)
			}
			
			if(!self.fulfilledQuestionnaires.isEmpty) {
				Text("error_reward_questionnaires_not_finished").padding(.vertical)
				let availableQuestionnaires = self.study.questionnaires.filter{(questionnaire: Questionnaire) in
					questionnaire.limitToGroup == 0 || questionnaire.limitToGroup == study.group || self.fulfilledQuestionnaires[KotlinLong(value: questionnaire.internalId)] ?? true != true
				}
				ForEach(availableQuestionnaires, id: \.internalId) { (questionnaire: Questionnaire) in
					HStack {
						Text(questionnaire.title)
						Spacer()
						if(self.fulfilledQuestionnaires[KotlinLong(value: questionnaire.internalId)] ?? true) as! Bool {
							Image(systemName: "checkmark.circle.fill").foregroundColor(Color.green)
						}
						else {
							Image(systemName: "xmark.circle.fill").foregroundColor(Color.red)
						}
					}
				}
			}
			if(study.enableRewardCalculation){
				VStack {
					if(!study.rewardCalculationInfo.isEmpty) {
						ScrollableHtmlTextView(html: study.rewardCalculationInfo)
					}
					Text(String(format: NSLocalizedString("reward_current_amount", comment: ""), study.getRewardAmount()))
				}
			}
			VStack {
				let untilActive = self.study.daysUntilRewardsAreActive()
				if untilActive > 0 {
					Text(String(format: NSLocalizedString("info_reward_not_active_yet", comment: ""), untilActive))
				}
				Spacer()
				Button(action: {
					self.showRequestConfirmation = true
				}) {
					Text("reward_code_request")
				}.disabled(!canRequest)
			}
			
			Spacer()
		}
		.padding()
		.alert(isPresented: $showRequestConfirmation) {
			Alert(
				title: Text("reward_code_request"),
				message: Text(study.enableRewardCalculation ? NSLocalizedString("reward_code_request_info_calculation", comment: "") : NSLocalizedString("reward_code_request_info", comment: "")),
				primaryButton: .default(Text("reward_code_request_action")) {
					self.requestCode()
				},
				secondaryButton: .cancel(Text("cancel"))
			)
		}
	}
	private func getLoadingView() -> some View {
		LoadingSpinner(isAnimating: .constant(true), style: .large)
	}
	private func getCodeView() -> some View {
		VStack {
			HStack {
				Text("colon_reward_code_header")
				Spacer()
				Button("what_for") {
					self.alertView = {
						Alert(title: Text("what_for"), message: Text(NSLocalizedString("reward_code_description", comment: "")), dismissButton: .default(Text("ok_")))
					}
					self.showAlert = true
				}
			}
			HStack {
				Button(action: {
					UIPasteboard.general.string = self.rewardCode
					self.appState.showTranslatedToast(String(format: NSLocalizedString("ios_info_copied_x_to_clipboard", comment: ""), self.rewardCode))
				}) {
					Text(self.rewardCode)
						.font(.system(size: 24))
					Image(systemName: "doc.on.clipboard")
						.foregroundColor(Color("Accent"))
				}
				.padding()
			}
			
			HStack {
				if(!study.contactEmail.isEmpty) {
					Spacer()
					Button(action: {
						var emailContent = self.study.rewardEmailContent.isEmpty ? NSLocalizedString("reward_code_content", comment: "") : self.study.rewardEmailContent
						emailContent = emailContent.replacingOccurrences(of: "[[CODE]]", with: self.rewardCode)
						if let url = URL(string: "mailto:\(self.study.contactEmail)?body=\(emailContent)".addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)!) {
							if #available(iOS 10.0, *) {
								UIApplication.shared.open(url)
							} else {
								UIApplication.shared.openURL(url)
							}
						}
					}) {
						VStack {
							Image(systemName: "envelope").frame(height: 30)
							Text("reward_code_send_email")
						}
					}
				}
				
				Spacer()
				Button(action: {
					self.showShareSheet = true
				}) {
					VStack {
						Image(systemName: "square.and.arrow.up").frame(height: 30)
						Text("reward_code_share")
					}
				}
				Spacer()
			}
			Spacer(minLength: 10.0)
			if(!study.rewardInstructions.isEmpty) {
				ScrollableHtmlTextView(html: study.rewardInstructions)
			}
			
			if(study.enableRewardCalculation) {
				VStack(alignment: .leading) {
					if(!study.rewardCalculationInfo.isEmpty) {
						ScrollableHtmlTextView(html: study.rewardCalculationInfo)
					}
					Text(String(format: NSLocalizedString("reward_final_amount", comment: ""), study.cachedRewardAmount))
				}
			}
				
		}
		.sheet(isPresented: $showShareSheet) {
			ShareSheet(activityItems: [self.rewardCode])
		}
		.alert(isPresented: self.$showAlert, content: self.alertView)
	}
	
	private func requestCode() {
		self.getCode = true
		self.fetchRewardCode()
	}
	
	private func fetchRewardCode() {
		error = ""
		self.study.getRewardCode(
			onError: { msg in
				error = msg
			},
			onSuccess: { rewardInfo in
				switch rewardInfo.errorCode {
				case Study.Companion().REWARD_SUCCESS:
					rewardCode = rewardInfo.code
				case Study.Companion().REWARD_ERROR_UNFULFILLED_REWARD_CONDITIONS:
					error = NSLocalizedString("error_reward_conditions_not_met", comment: "")
					fulfilledQuestionnaires = rewardInfo.fulfilledQuestionnaires
				case Study.Companion().REWARD_ERROR_ALREADY_GENERATED:
					error = NSLocalizedString("error_already_generated", comment: "")
				default:
					error = rewardInfo.errorMessage
				}
			}
		)
	}
	
	var body: some View {
		ScrollView{
			HStack {
				let untilActive = self.study.daysUntilRewardsAreActive()
				if(!self.study.enableRewardSystem || untilActive != 0) {
					self.getDefaultView(error: String(format: NSLocalizedString("info_reward_is_not_active_yet", comment: ""), untilActive))
				}
				else if(self.study.enableRewardCalculation) {
					if(!self.error.isEmpty){
						self.getDefaultView(error: self.error)
					} else if(!self.rewardCode.isEmpty) {
						self.getCodeView()
					} else if(self.getCode){
						self.getLoadingView()
					} else {
						self.getDefaultView(error: "")
					}
				}
				else {
					if(!self.error.isEmpty) {
						self.getDefaultView(error: self.error)
					}
					else if(rewardCode.isEmpty) {
						self.getLoadingView()
					}
					else {
						self.getCodeView()
					}
				}
			}.padding()
		}
		.padding()
		.onAppear {
			getCode = study.hasCachedRewardCode()
			if getCode {
				fetchRewardCode()
			} else {
				fulfilledQuestionnaires = (study.getRewardFulfillmentLocal() as? [KotlinLong : KotlinBoolean]) ?? [:]
			}
		}
	}
}
